### DISCUSSION TITLE
Model as parameter

### DISCUSSION THREAD

**Discussion by UesleyCarossi** [adonisjs/core#4507] [Ideas]
### **Current behavior:**
When using dependency inversion with BaseModel, LucidModel built-in properties and methods (like find, findBy, and query) become inaccessible. This makes it difficult to leverage model functionality when passing them as parameters to other functions or classes.

![Screenshot from 2024-04-11 14-13-28](https://github.com/adonisjs/core/assets/43580005/d8807ecf-a1ae-4ada-85fd-862618aa300a)

### **Desired behavior:**
We would like the ability to pass Lucid models as parameters while still maintaining access to their built-in properties and methods. This would allow for a more flexible and reusable approach to dependency injection with Lucid models.

### **Benefits:**
- Improved code maintainability and testability through loose coupling.
- Increased code reusability by enabling models to be used in various contexts.
- Better adherence to dependency inversion principles.

**Comment by thetutlage** [comment]
The `findBy` is a static method and you are passing an instance of `User` model to the class. So it's technically not possible. If needed you must inject the class itself, not its instance.

**Reply by UesleyCarossi** [reply]
That way my code is less reusable and create a strong dependency relationship with framework, in example, is required to use the same framework to create tests.

**Reply by thetutlage** [reply]
I am not even sure what the complain here is.

- Static methods are not instance methods.
- When you inject an instance of a class, you have to access to instance properties and not static methods/properties.

All these are JavaScript concerns and has nothing to do with AdonisJS.

You can use Repository pattern and abstract database operations in there.
